### DISCUSSION TITLE
Craft 5 - Licencing with Top Level Domain

### DISCUSSION THREAD

**Discussion by magnessjo** [craftcms/cms#14550] [Ideas] [answered]
After Craft 4.8.1 and Craft 5 beta 4, the new licensing rules came into effect with a warning page and a timer. 

I assume that Pixel&Tonic is doing a regex check for a top-level domain. 

I request that a check be included on that string to check the existence of "localhost."  Localhost is a commonly used label for URLs in a local environment, but for many developers, the URL doesn't end with *.localhost. 

For example, I've used *.localhost.dev since before Google introduced .dev as a top-level domain.

**Comment by brandonkelly** [comment, marked as answer]
We just updated our license checks to treat `localhost.dev` as a development domain.

Run the following command to refresh Craft’s internal cache, and then it will know it’s OK to run trials the next time it pings our API.

```sh
php craft clear-caches/data
```

**Reply by Anubarak** [reply]
Is there any way to include `.dev` domains alltogether?
All our internal domains end with `2ndlab.dev` and all our projects have license issues (they had them before already but we just ignored the "this license belongs to >insert live url<" warning)

For example:

https://craft48license.2ndlab.dev/admin/

Since waiting more and more every day for every project will be tedious to say at least we would have to overwrite your check if there is nothing else we could do.

Even this URL complains
https://testinstall.2ndlab.dev/admin/ (note we installed Pro as Trial)
![image](https://github.com/craftcms/cms/assets/6122122/c2bf9718-cfa3-41ce-b304-a231fa878df5)

**Reply by brandonkelly** [reply]
@Anubarak `.dev` has been a publicly-registrable TLD since 2019, so no, we cannot treat all `.dev` domains as for development/staging, any more than we would treat all `.com` or `.net` domains that way.

> (they had them before already but we just ignored the "this license belongs to >insert live url<" warning)

The fact that the alerts were easy to ignore is exactly why we decided to take additional measures. Unfortunately we have no way of distinguishing between issues that are being ignored with the intent to fix, vs issues being ignored because the license is delinquent.

> Since waiting more and more every day for every project will be tedious to say at least we would have to overwrite your check if there is nothing else we could do.

Obviously I can’t endorse that approach and need to point out that tampering with Craft’s license enforcement code is strictly against the license agreement.

There is another option, which is you can start using a sanctioned [development-sounding subdomain](https://craftcms.com/knowledge-base/how-craft-license-enforcement-works), which will allow Craft to run trials just like a local testing domain like `.test`, `.site`, etc. For example, you could use `projectname-staging.2ndlab.dev`.

Or, if `2ndlab.dev` is **exclusively** used for staging purposes, you can have https://2ndlab.dev redirect to your public site, and we can add the domain to our list of staging-only domains, which will let you continue using `projectname.2ndlab.dev` without any licensing issues.

**Reply by Anubarak** [reply]
I see, thank you for your fast reply.

> The fact that the alerts were easy to ignore is exactly why we decided to take additional measures

I can totally see that and I really get why you did it. Sorry if my first question sounded a little bit too rude.
I really didn't want to offend you, I was just looking for a way to continue our local development without having to wait every day. Even projects that already have a correct license and are already live do complain so I thought there was no real way to fix it (besides changing the URL). I really didn't want to bypass the requirement for your license - just the requirement to buy two licenses (one for development and one for production which was not your intention)

> There is another option, which is you can start using a sanctioned [development-sounding subdomain](https://craftcms.com/knowledge-base/how-craft-license-enforcement-works)

I already read this link several times but we thought having a `.dev` domain would be enough. Now it's clear, thank you again.

> Or, if 2ndlab.dev is exclusively used for staging purposes, you can have https://2ndlab.dev/ redirect to your public site

https://2ndlab.dev/ points to our development server and is **exclusively** used for development. We all work on this dev server so no one has to work on their own machines. The domain https://2ndlab.dev/ itself does actually nothing, it just shows some info. I could output a 404 error page as well.

All our projects in development have their own subdomain like `https://project-name.2ndlab.dev` or `https://project-name.2ndred.dev/` (we use both, 2ndlab and 2ndred).
Could you please explain where https://2ndlab.dev/ should redirect to? Maybe our public site in terms of our official website https://www.secondred.de/ or to some random public site from a project that we made?
We would be very happy and grateful if you could include `2ndlab.dev` in your checks rather than adding a `-staging` to every subdomain.

**Reply by brandonkelly** [reply]
Alright, I just went ahead and added `2ndlab.dev` to our dev domain list. Clear data caches and then Craft will stop showing licensing issues for trials.

**Reply by Anubarak** [reply]
Absolutely awesome, thank you so much.
It works perfectly
